### PR TITLE
MMI-antiLM decoding

### PR DESCRIPTION
**Pull request by AADeLucia**
# What does this PR do?

Implements **Maximum Mutual Information anti-Language Model objective (MMI anti-LM)** decoding from [A Diversity-Promoting Objective Function for Neural Conversation Models](https://www.aclweb.org/anthology/N16-1014.pdf).

### CODE REVIEW

**Comment by sshleifer** [comment]
This technique looks really cool! Unfortunately running `generate` with two models will break lots of our assumptions, so maybe you could write a standalone:

```python
def generate_antimlm(model, lm_model, **kwargs):
    return generated_ids
```

,put it in `examples/anti_mlm_generation/`, and add a test that it runs with `sshleifer/tiny-gpt2`, for example?

Does that make sense to you @patrickvonplaten ?

File: src/transformers/generation_utils.py [hunk 1]
```diff
@@ -526,6 +529,18 @@ def _generate_no_beam_search(
...
+
+        # Initialize MMI antiLM settings
+        if antilm_params is None:
+            print("Must define antilm_params")
+            quit()
+        threshold_g = antilm_params["length_thresh"] + cur_len  # cur_len starts at the length of input
+        alpha = antilm_params["alpha"]
+        lm_model = antilm_params["lm_model"]
+        bsz = input_ids.size(0)
+        input_ids2 = torch.LongTensor(bsz,1).fill_(50256).to(input_ids.device)
```
Comment (sshleifer) [inline src/transformers/generation_utils.py:542]: is `50256` some symbolic token like `pad_token_id`?

**Review by sshleifer — CHANGES_REQUESTED** [review]
See comment, I think this fits better in `examples/`

File: src/transformers/generation_utils.py [hunk 2]
```diff
@@ -526,6 +529,18 @@ def _generate_no_beam_search(
         sent_lengths = input_ids.new(batch_size).fill_(max_length)
 
         past = None
+        past2 = None
+
+        # Initialize MMI antiLM settings
+        if antilm_params is None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_utils.py:535]: This is `None` by default, right? So by default this branch would be hit and `generate()` would `quit()`?

**Comment by patrickvonplaten** [comment]
Hey @AADeLucia - thanks for your PR! The `generate()` function is a very central part of the library and thus we have to be super careful when implementing new features. I agree with @sshleifer that MMI-antiLM decoding probably fits better in an example (has its own `generate()` function in an example file) to begin with - would that be ok for you?

File: src/transformers/generation_utils.py (hunk 1 above)
Comment (AADeLucia) [reply src/transformers/generation_utils.py:542]: Yes, sorry that was hard-coded for GPT-2.

File: src/transformers/generation_utils.py (hunk 2 above)
Comment (AADeLucia) [reply src/transformers/generation_utils.py:535]: That line is just error-handling for my research code.

**Comment by AADeLucia** [comment]
> Hey @AADeLucia - thanks for your PR! The `generate()` function is a very central part of the library and thus we have to be super careful when implementing new features. I agree with @sshleifer that MMI-antiLM decoding probably fits better in an example (has its own `generate()` function in an example file) to begin with - would that be ok for you?

Thank you so much for your quick responses! Yes, putting it as it's own example is okay with me.

**Pull request closed without merging**